### ISSUE TITLE
Android Client started throwing "ca.uhn.fhir.parser.DataFormatException" after upgrading to 1.6

### ISSUE THREAD

**Issue by viralypatel** [jamesagnew/hapi-fhir#422]
It was all working fine till i was using version 1.5. Yesterday upgraded the dependency to 1.6 and started getting the `ca.uhn.fhir.parser.DataFormatException` for all resources that has a date field.

Dependency added as follows:

```
compile(group: 'ca.uhn.hapi.fhir', name: 'hapi-fhir-android', version: '1.6', classifier: 'dstu2') {
        exclude module: 'javax.servlet-api'
        exclude module: 'hapi-fhir-base'
    }
```

**Exception stackTrace**

```
08-11 14:28:41.186 28961-28961/com.myapp.playground W/System.err: Caused by: ca.uhn.fhir.parser.DataFormatException: Invalid date/time format: "2012-09-17T24:12:00"
08-11 14:28:41.187 28961-28961/com.myapp.playground W/System.err:     at ca.uhn.fhir.model.primitive.BaseDateTimeDt.throwBadDateFormat(BaseDateTimeDt.java:464)
08-11 14:28:41.187 28961-28961/com.myapp.playground W/System.err:     at ca.uhn.fhir.model.primitive.BaseDateTimeDt.parseInt(BaseDateTimeDt.java:370)
08-11 14:28:41.187 28961-28961/com.myapp.playground W/System.err:     at ca.uhn.fhir.model.primitive.BaseDateTimeDt.parse(BaseDateTimeDt.java:306)
08-11 14:28:41.188 28961-28961/com.myapp.playground W/System.err:     at ca.uhn.fhir.model.primitive.BaseDateTimeDt.parse(BaseDateTimeDt.java:39)
08-11 14:28:41.188 28961-28961/com.myapp.playground W/System.err:     at ca.uhn.fhir.model.api.BasePrimitive.setValueAsString(BasePrimitive.java:112)
08-11 14:28:41.188 28961-28961/com.myapp.playground W/System.err:     at ca.uhn.fhir.model.primitive.BaseDateTimeDt.setValueAsString(BaseDateTimeDt.java:460)
08-11 14:28:41.188 28961-28961/com.myapp.playground W/System.err:     at ca.uhn.fhir.parser.ParserState$PrimitiveState.attributeValue(ParserState.java:2355)
08-11 14:28:41.188 28961-28961/com.myapp.playground W/System.err:     at ca.uhn.fhir.parser.ParserState.attributeValue(ParserState.java:116)
08-11 14:28:41.188 28961-28961/com.myapp.playground W/System.err:     at ca.uhn.fhir.parser.JsonParser.parseChildren(JsonParser.java:1301)
08-11 14:28:41.188 28961-28961/com.myapp.playground W/System.err:     at ca.uhn.fhir.parser.JsonParser.parseChildren(JsonParser.java:1226)
08-11 14:28:41.189 28961-28961/com.myapp.playground W/System.err:     at ca.uhn.fhir.parser.JsonParser.parseChildren(JsonParser.java:1291)
08-11 14:28:41.189 28961-28961/com.myapp.playground W/System.err:     at ca.uhn.fhir.parser.JsonParser.parseChildren(JsonParser.java:1226)
08-11 14:28:41.189 28961-28961/com.myapp.playground W/System.err:     at ca.uhn.fhir.parser.JsonParser.parseChildren(JsonParser.java:1291)
08-11 14:28:41.189 28961-28961/com.myapp.playground W/System.err:     at ca.uhn.fhir.parser.JsonParser.parseChildren(JsonParser.java:1274)
08-11 14:28:41.189 28961-28961/com.myapp.playground W/System.err:     at ca.uhn.fhir.parser.JsonParser.parseChildren(JsonParser.java:1226)
08-11 14:28:41.189 28961-28961/com.myapp.playground W/System.err:     at ca.uhn.fhir.parser.JsonParser.doParseResource(JsonParser.java:231)
08-11 14:28:41.189 28961-28961/com.myapp.playground W/System.err:     at ca.uhn.fhir.parser.BaseParser.parseResource(BaseParser.java:604)
08-11 14:28:41.189 28961-28961/com.myapp.playground W/System.err:     at ca.uhn.fhir.rest.client.BaseClient$ResourceResponseHandler.invokeClient(BaseClient.java:534)
08-11 14:28:41.190 28961-28961/com.myapp.playground W/System.err:     at ca.uhn.fhir.rest.client.BaseClient$ResourceResponseHandler.invokeClient(BaseClient.java:486)
08-11 14:28:41.190 28961-28961/com.myapp.playground W/System.err:     at ca.uhn.fhir.rest.client.BaseClient.invokeClient(BaseClient.java:352)
08-11 14:28:41.190 28961-28961/com.myapp.playground W/System.err: ... 12 more
```

Is there anything additional required to be done when upgrading to version 1.6? Anyone who faced the same and resolved it?

**Comment by sjanic** [comment]
I'm having a similar issue. I have made a reproducible example where I can provoke the exception. Given the following profiles:
[CustomProcedureRequest.txt](https://github.com/jamesagnew/hapi-fhir/files/413173/CustomProcedureRequest.txt)

[CustomTimingDt.txt](https://github.com/jamesagnew/hapi-fhir/files/413174/CustomTimingDt.txt)

And the following activity:
[CustomProcedureRequestActivity.txt](https://github.com/jamesagnew/hapi-fhir/files/413177/CustomProcedureRequestActivity.txt)

I get the same exception, when I press the button in my activity. It seems like the android-fhir-library 1.6 is having some troubles parsing the CustomTimingDt datatype in the CustomProcedureRequest.

**Comment by sjanic** [comment]
I just tried to run my reproducible example in the normal HAPI client as well, and I get the same bug so I might suspect that it isn't an Android specific bug.

As to what the bug might be, I have an idea:
I debugged the ModelScanner.java, which from what I could tell is responsible for loading the FhirContext with the build-in and custom Resource and Datatypes. In the ModelScanner it splits the Resources and Datatypes to different RuntimeDefinitions it referrences later for parsing resources to XML, JSON etc.

When the RuntimeDefinitions are defined, a sealAndInitialize method is called to fill out the definition with the necessary Resource or Datatype information so it can be used later by the parsers.
The HAPI client encounters my CustomProcedureRequest, and defines its RuntimeDefinitions. When it gets to the scheduled field, the ModelScanner isn't aware of the scheduled's datatype CustomTimingDt, because we haven't encountered it yet, and cannot define it at that time. But since the RuntimeDefinition for scheduled is sealed afterwards, it can never update its definition, and the parser will fail with the above mentioned exception.

It seems that in RuntimeChildDeclaredExtensionDefinition.java we are aware of this:
```
@Override
void sealAndInitialize(FhirContext theContext, Map<Class<? extends IBase>, BaseRuntimeElementDefinition<?>> theClassToElementDefinitions) {
	myUrlToChildExtension = new HashMap<String, RuntimeChildDeclaredExtensionDefinition>();
		
	BaseRuntimeElementDefinition<?> elementDef = theClassToElementDefinitions.get(myChildType);
		
	/*
	 * This will happen for any type that isn't defined in the base set of 
	 * built-in types, e.g. custom structures or custom extensions
	 */
	if (elementDef == null) {
		elementDef = theContext.getElementDefinition(myChildType);
	}

...
```
But not in RuntimeChildChoiceDefinition, which scheduled in CustomProcedureRequest gets defined as. @jamesagnew Could it be that some of the RuntimeDefinitions simply are missing this guard?
